### PR TITLE
refactor(state): decode raft proposals once at the applier boundary

### PR DESCRIPTION
**Pull request by gfyrag**
## Summary

The FSM apply pipeline used to `UnmarshalVT` the same `entry.Data` up to **three times** per Raft entry:

1. `applier.findCheckpointBoundary` — scan to find the checkpoint pre-split boundary
2. `state.ValidateCheckpointEntryPositions` — defensive validator called at the top of `PrepareEntries`
3. `state.Machine.PrepareEntries` main loop — the actual apply decode

For a batch of `N` entries that meant ~`3N` decodes of the same payload. This PR moves the decode to the **applier boundary** so it happens once per entry, and threads the decoded form through every downstream stage.

## What changed

- **New** `internal/infra/state/decoded_entry.go`: `DecodedEntry` (entry + pre-decoded `*raftcmdpb.Proposal`), `DecodeEntries`, `ReleaseDecodedEntries`, `DecodedEntryRequiresCheckpoint`.
- **State**: `PrepareEntries(...raftpb.Entry)` is now a thin wrapper that decodes → delegates → releases. The hot path core is `PrepareDecodedEntries(...DecodedEntry)` — no `UnmarshalVT` in the apply loop. Added `ValidateCheckpointEntryPositionsDecoded` (no-unmarshal sibling kept for parity with the raw variant).
- **Applier**: `applyEntriesToFSM` and `applyReplayEntries` decode once at the top (`defer state.ReleaseDecodedEntries(decoded)`), then pass the `[]state.DecodedEntry` through `findCheckpointBoundaryDecoded` → `applyEntriesPipelined` / `applyEntriesAndResolveCommands` → `PrepareDecodedEntries`. Helper `rawEntriesFromDecoded` for the few sites that need `[]raftpb.Entry` back (spool tail append, checkpoint maintenance handoff).

## Cost per batch of N entries

| Stage | Before | After |
|---|---|---|
| `findCheckpointBoundary` | 1 unmarshal / entry | 0 |
| `ValidateCheckpointEntryPositions` | 1 unmarshal / (N-1) | 0 |
| `PrepareEntries` apply loop | 1 unmarshal / entry | 0 |
| `DecodeEntries` at applier boundary | — | **1 unmarshal / entry** |
| **Total** | **~3N** | **1N** |

## Lifetime and safety

The applier owns each `*raftcmdpb.Proposal`'s lifetime via `defer state.ReleaseDecodedEntries(decoded)`. The returned `*PreparedBatch` retains no pointer into the proposal — every per-entry effect is captured by value into `ApplyResult` during apply. This matches the invariant already implicit in the previous single-`cmd` VT-pool recycling pattern.

## Backward compatibility

Public signatures preserved as wrappers (`PrepareEntries`, `ApplyEntries`, `findCheckpointBoundary`, `ValidateCheckpointEntryPositions`) — no test file needed to change. ~30+ call sites across `internal/infra/state/*_test.go` and `internal/infra/node/checkpoint_boundary_test.go` continue to work as-is.

## Test plan

- [x] `go build ./...`
- [x] `go test ./internal/...` (state + node + all other internals)
- [x] `golangci-lint run --new-from-rev=HEAD~ ./internal/infra/state/... ./internal/infra/node/...` — 0 issues
- [ ] CI green on this PR
- [ ] Antithesis run (optional sanity check given hot-path nature)

### CODE REVIEW

File: internal/infra/state/decoded_entry.go [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+// DecodeEntries unmarshals every normal entry's Data into a pooled
+// *raftcmdpb.Proposal. Non-normal entries (raftpb.EntryConfChange*) and
+// entries with empty Data get a nil Proposal so the slice keeps positional
+// alignment with the source entries.
+//
+// On error, any proposals already decoded are returned to the pool and a
+// nil slice is returned so callers need not call ReleaseDecodedEntries
+// themselves.
+func DecodeEntries(entries []raftpb.Entry) ([]DecodedEntry, error) {
```
Comment (paul-nicolas) [inline internal/infra/state/decoded_entry.go:39]: [Medium] New state-package functions ship with no unit tests

`DecodeEntries` (here), `ReleaseDecodedEntries`, and `ValidateCheckpointEntryPositionsDecoded` (`checkpoint_orders.go:125`) are added with **zero** direct test coverage — the diff adds no `_test.go` files. The decoded checkpoint-boundary scan is only exercised indirectly via the retained `findCheckpointBoundary` wrapper, and the two paths most prone to silent regression are untested:

- **`DecodeEntries` error / partial-release path** (`UnmarshalVT` failure → `ReleaseDecodedEntries(decoded[:i])` → nil slice): a future edit here could double-free or leak pooled `*raftcmdpb.Proposal`s with no test to catch it. This is pool-accounting code on the FSM hot path — exactly where a regression is most costly.
- **`ReleaseDecodedEntries`** idempotency / nil-out behavior.

Per CLAUDE.md ("Write tests for all new features"), add `internal/infra/state/decoded_entry_test.go` covering: (a) a mixed slice (normal + conf-change + empty-data) producing correctly positioned nil `Proposal`s; (b) a malformed `entry.Data` returning a nil slice with the VT pool left balanced (no double-free); (c) `ValidateCheckpointEntryPositionsDecoded` parity cases mirroring `TestValidateCheckpointEntryPositions` (trigger-middle, trigger-last, nil-Proposal-then-trigger-last, empty slice).

**Review by paul-nicolas — COMMENTED** [review]
## Multi-model PR review (Claude + Codex)

**Note:** the Codex CLI was unavailable in this environment (`omc ask codex` failed: advisor script not found), so this is effectively a Claude-only review — a second independent Claude reviewer plus the orchestrator's own line-by-line verification against the PR head (`4c6d5b1`).

### Verdict: COMMENT — no blocking issues

This decode-once hot-path refactor is **correct**. All five scrutinized risk areas were independently verified against the actual PR-head code:

- **VT-pool lifetime / use-after-release:** the async committer (`CommitPreparedBatch`) reads **no** `*raftcmdpb.Proposal` pointer; `PreparedBatch`/`ApplyResult` retain none (every per-entry effect captured by value). Decisive: the old code returned each proposal to the pool *during* the prepare loop — before `submitAsyncCommit` ran — so the new code (which holds proposals until `applyEntriesToFSM`/`applyReplayEntries` returns via `defer ReleaseDecodedEntries`) keeps them strictly **longer**. No new use-after-free is possible.
- **`sweepBelowTerm` → inline `maxTerm` loop:** semantically identical (same term, same `> 0` guard, same `FailFuturesBelowTerm` call).
- **`ValidateCheckpointEntryPositionsDecoded`:** parity with the raw variant — nil-Proposal handling, `range`-over-int loop bounds (exactly the non-last entries), and error semantics all match.
- **`DecodeEntries` error path:** on `UnmarshalVT` failure the failing `cmd` is pooled, `ReleaseDecodedEntries(decoded[:i])` releases only `[0,i)`, `decoded[i].Proposal` is still nil — no leak, no double-release.
- **CLAUDE.md invariants:** FSM stays deterministic; the new nil-Proposal branch fails loudly (`assert.Unreachable` + `batch.Cancel()` + `return fmt.Errorf("invariant: ...")`) per invariant #7; no Pebble reads added on the hot path. No API/CLI/OpenAPI/proto surface touched, so no documentation drift.

### Findings
1 Medium (posted inline), 2 Minor (listed below). A low-confidence `*raftcmdpb.VolumePair`-provenance question was investigated and **dropped**: `sentinelUpdates` come from the `WriteSet` overlay (FSM-computed deltas), not the wire proposal, and that provenance is unchanged old-vs-new — no regression is possible.

### Minor / Nit (not posted inline)
- `internal/infra/state/machine.go:667,677` — stale `"PrepareEntries batch"` wording in the `assert.Unreachable` label and the error string; the logic now lives in `PrepareDecodedEntries`. Rename both for log/grep consistency.
- `internal/infra/node/applier.go:173` — `rawEntriesFromDecoded` allocates a fresh `[]raftpb.Entry` per checkpoint/tail split. This is off the steady-state hot path (only when a checkpoint trigger fires or a tail must be spooled), so the cost is negligible. Noted as a conscious choice, not an oversight — no change required.

File: internal/infra/state/decoded_entry.go (hunk 1 above)
Comment (NumaryBot) [inline internal/infra/state/decoded_entry.go:39]: 🟡 **[minor] No unit tests for new state-package decode/release functions**

DecodeEntries, ReleaseDecodedEntries, and ValidateCheckpointEntryPositionsDecoded are added with no direct test coverage. The error/partial-release path in DecodeEntries (UnmarshalVT failure → ReleaseDecodedEntries(decoded[:i])) is pool-accounting code on the FSM hot path and is only exercised indirectly. A future edit could introduce a double-free or leak of pooled *raftcmdpb.Proposal objects with no test to catch it.

**Suggestion:** Add internal/infra/state/decoded_entry_test.go covering: (a) a mixed slice (normal + conf-change + empty-data entries) producing correctly-positioned nil Proposal values; (b) a malformed entry.Data returning a nil slice with the VT pool left balanced; (c) ReleaseDecodedEntries idempotency/nil-out behaviour; (d) ValidateCheckpointEntryPositionsDecoded boundary conditions.

**Review by NumaryBot — COMMENTED** [review]
NumaryBot posted 1 new inline finding.

Summary: https://github.com/formancehq/ledger/pull/1453#issuecomment-4841693041

File: internal/infra/state/decoded_entry.go (hunk 1 above)
Comment (gfyrag) [reply internal/infra/state/decoded_entry.go:39]: Addressed in 0ee7765e0 — added `internal/infra/state/decoded_entry_test.go` (280 lines, +47 sub-tests) covering exactly the cases you flagged:

- **`DecodeEntries`**: nil/empty input, mixed slice (normal + conf-change + empty-data) with positional-alignment + nil-Proposal assertions, malformed payload at trailing and first positions (returns error + nil slice, the partial-release happens internally and is verified via the double-release idempotency test below — a leak would surface as a pool corruption in subsequent runs).
- **`ReleaseDecodedEntries`**: nil/empty no-op, and a double-release case asserting every `decoded[i].Proposal` is nilled out so a second call is safe.
- **`DecodedEntryRequiresCheckpoint`**: nil / apply-only / checkpoint-trigger.
- **`ValidateCheckpointEntryPositionsDecoded`**: full parity with `TestValidateCheckpointEntryPositions` — empty, no-trigger, trigger-last/first/middle/single, conf-change + empty-data skipped, plus the `nil-proposal entry then trigger last` case you specifically called out.

Also picked up the `machine.go:667,677` minor in the same commit ("PrepareEntries batch" → "PrepareDecodedEntries batch" in the `assert.Unreachable` label and the error string, for grep/observability consistency).

Skipping the `rawEntriesFromDecoded` minor per your own note ("no change required").

Comment (NumaryBot) [inline internal/infra/state/decoded_entry.go:39]: 🟡 **[minor] No unit tests for new decode/release helper functions**
<sub>reported by paul-nicolas, NumaryBot, codex</sub>

DecodeEntries, ReleaseDecodedEntries, and ValidateCheckpointEntryPositionsDecoded are added with no direct test coverage. The error/partial-release path in DecodeEntries (UnmarshalVT failure → ReleaseDecodedEntries(decoded[:i])) is pool-accounting code on the FSM hot path and is only exercised indirectly via retained wrappers. A future edit could introduce a double-free or leak of pooled *raftcmdpb.Proposal objects with no test to catch it.

**Suggestion:** Add internal/infra/state/decoded_entry_test.go covering: (a) a mixed slice (normal + conf-change + empty-data entries) producing correctly-positioned nil Proposal values; (b) a malformed entry.Data returning a nil slice with the VT pool left balanced (no double-free); (c) ReleaseDecodedEntries idempotency/nil-out behaviour; (d) ValidateCheckpointEntryPositionsDecoded boundary conditions.

**Review by NumaryBot — COMMENTED** [review]
NumaryBot posted 1 new inline finding.

Summary: https://github.com/formancehq/ledger/pull/1453#issuecomment-4841693041

File: internal/infra/state/decoded_entry.go (hunk 1 above)
Comment (gfyrag) [reply internal/infra/state/decoded_entry.go:39]: Stale finding — re-scanned on an older snapshot. The tests landed in `0ee7765e0` (current PR head) as `internal/infra/state/decoded_entry_test.go` (+280 lines, 47 sub-tests).

Specifically covering the four cases listed:

- **(a)** mixed slice (normal + conf-change + empty-data) producing correctly-positioned nil `Proposal`s → `TestDecodeEntries/mixed:_normal,_conf-change,_empty-data`
- **(b)** malformed `entry.Data` returning a nil slice with the VT pool left balanced → `TestDecodeEntries/malformed_payload_returns_error_and_nil_slice` and `.../malformed_payload_at_first_position`; the no-double-free is verified by `TestReleaseDecodedEntries/nils_out_Proposal_pointers_so_double-release_is_safe` (asserts `Proposal == nil` post-release and that a second release is a no-op)
- **(c)** `ReleaseDecodedEntries` idempotency / nil-out → same test as above + `nil_slice_is_a_no-op` + `empty_slice_is_a_no-op`
- **(d)** `ValidateCheckpointEntryPositionsDecoded` boundary conditions → `TestValidateCheckpointEntryPositionsDecoded` with full parity against `TestValidateCheckpointEntryPositions` (empty, no-trigger, trigger-last/first/middle/single, conf-change + empty-data skipped, nil-proposal-then-trigger-last)

**Comment by gfyrag** [comment]
@Azorlogh ping pour review 🙏

@paul-nicolas — j'ai adressé les findings de ton tour précédent (Medium tests dans `0ee7765e0`, Minor 1 string rename même commit ; Minor 2 skippé per ta note). PR rebasée sur `release/v3.0` (CI verte sauf `codecov/patch` qui est structurel — détail dans le thread d'origine). Si tu peux re-jeter un œil quand tu as un moment !

**Review by paul-nicolas — APPROVED** [review]
